packages/docusaurus-theme-common/src/utils/codeBlockUtils.ts

```diff
@@ -9,8 +9,36 @@ import type {CSSProperties} from 'react';
 import rangeParser from 'parse-numeric-range';
 import type {PrismTheme, PrismThemeEntry} from 'prism-react-renderer';
 
-const codeBlockTitleRegex = /title=(?<quote>["'])(?<title>.*?)\1/;
-const metastringLinesRangeRegex = /\{(?<range>[\d,-]+)\}/;
+// note: regexp/no-useless-non-capturing-group is a false positive
+// the group is required or it breaks the correct alternation of
+// <quote><stringValue><quote> | <rawValue>
+const optionRegex =
+  // eslint-disable-next-line regexp/no-useless-non-capturing-group
+  /(?<key>\w+)(?:=(?:(?:(?<quote>["'])(?<stringValue>.*?)\k<quote>)|(?<rawValue>\S*)))?/g;
+const metastringLinesRangeRegex = /\{(?<range>[\d,-]+)\}/g;
+const highlightOptionKey = 'highlight';
+
+/**
+ * The supported types for {@link CodeBlockMeta.options} values.
+ */
+export type CodeMetaOptionValue = string | boolean | number;
+
+/**
+ * Any options as specified by the user in the "metastring" of codeblocks.
+ */
+export interface CodeBlockMeta {
+  /**
+   * The highlighted lines, 0-indexed. e.g. `{ 0: ["highlight", "sample"] }`
+   * means the 1st line should have `highlight` and `sample` as class names.
+   */
+  readonly lineClassNames: {[lineIndex: number]: string[]};
+
+  /**
+   * The parsed options, key converted to lowercase.
+   * e.g. `"title" => "file.js", "showlinenumbers" => true`
+   */
+  readonly options: {[key: string]: CodeMetaOptionValue};
```

Comment:
This is not ideal, not typesafe and not extendable

Let's use a `CodeBlockMetaOptions` interface and leverage TS declaration merging to register extra options (for example the live codeblock can register its `noInline` option)

Reply:
Easy to change, will do. This was initially inspired by the general code style in this area (e.g. `lineClassNames`). -> Also see remark regarding case insenstivity below which could influence the design of `CodeBlockMetaOptions`.

Reply:
> This was initially inspired by the general code style in this area (e.g. `lineClassNames`). 

No idea what you mean 😅 

> Also see remark regarding case insenstivity below which could influence the design of `CodeBlockMetaOptions`.

Even if we implement case insensitiveness, we still want `noInline` as a typesafe option attribute

Reply:
Concrete type definitions and documentation considered in https://github.com/facebook/docusaurus/pull/11017